File: DotNet/OSG/Glue/Factory.cpp

///////////////////////////////////////////////////////////////////////////////
//
//  Copyright (c) 2006, Adam Kubach
//  All rights reserved.
//  BSD License: http://www.opensource.org/licenses/bsd-license.html
//
///////////////////////////////////////////////////////////////////////////////

#include "Factory.h"

using namespace CadKit::OSG::Glue;

Factory*  Factory::_instance( 0x0 );


Factory& Factory::instance()
{
  if( 0x0 == _instance )
    _instance = new Factory();

  return *_instance;
}


Factory::Factory() : _map()
{
}

void Factory::create ( unsigned int key )
{
  _map[key] = new osg::Node;
}

osg::Node* Factory::node ( unsigned int key )
{
  NodeMap::const_iterator iter = _map.find( key );
  if( iter != _map.end() )
    return iter->second.get();

  return 0x0;
}

void Factory::node ( unsigned int key, osg::Node* node )
{
  _map[key] = node;
}

void Factory::remove ( unsigned int key )
{
  NodeMap::iterator iter = _map.find( key );
  if( iter != _map.end() )
    _map.erase( iter );
}

File: branches/usul_mutex_test/Legacy/DotNet/Plugins/Options/Ossim/Factory.cs

///////////////////////////////////////////////////////////////////////////////
//
//  Copyright (c) 2006, Perry L Miller IV
//  All rights reserved.
//  BSD License: http://www.opensource.org/licenses/bsd-license.html
//
///////////////////////////////////////////////////////////////////////////////

namespace CadKit.Plugins
{
  class Factory :
    CadKit.Referenced.Base,
    CadKit.Interfaces.IClassFactory
  {
    /// <summary>
    /// Data members.
    /// </summary>
    class TypeMap : System.Collections.Generic.Dictionary<string,System.Type> { }
    private TypeMap _types = new TypeMap();
    private object _mutex = new object();

    /// <summary>
    /// Constructor.
    /// </summary>
    public Factory()
    {
      this._addType("CadKit.Interfaces.IPlugin", "CadKit.Plugins.Options.Ossim.Component");
    }

    /// <summary>
    /// Add a type.
    /// </summary>
    private void _addType ( string key, string name )
    {
      lock (_mutex)
      {
        System.Type type = System.Type.GetType(name);
        _types[key] = type;
      }
    }

    /// <summary>
    /// Create the instance.
    /// </summary>
    object CadKit.Interfaces.IClassFactory.createInstance(string key)
    {
      try
      {
        lock (_mutex)
        {
          System.Type type = _types[key];
          string name = type.ToString();
          return System.Reflection.Assembly.GetExecutingAssembly().CreateInstance(name);
        }
      }
      catch (System.Collections.Generic.KeyNotFoundException)
      {
        return null;
      }
    }
  }
}
